[aarch64] use crc32 instructions when available

sebpop opened the pull request:
The time goes from 0.175s down to 0.007s on a Graviton A1 instance on the
following benchmark:
```
function simple_crc32() {
  $a = "foo";
  for ($i = 0; $i < 10000; $i++) {
      crc32($a);
      $a .= "o";
  }
}
```

sebpop commented:
I'm not sure what you are asking for, could you please be more specific?

The benchmark is a synthetic one that I run with the timing code from Zend/bench.php.

peter279k commented:
@sebpop, thank you for your reply.

I want to know where the reference presents about this performance issue.

sebpop commented:
Is there anything that needs to be addressed in this change, or is it ok to commit?
Thanks!

dstogov commented on ext/standard/crc32.c:
```diff
@@ -28,13 +55,60 @@ PHP_NAMED_FUNCTION(php_if_crc32)
 	crc = crcinit^0xFFFFFFFF;
 
+#if defined(__aarch64__)
+	if (has_crc32_insn()) {
```
It probably, may make sense to reuse ZEND_INTRIN_... mechanism to use link-time ifunc resolver.

sebpop replied:
Thanks for the feedback: I will amend the patch and resubmit for review.

nikic replied:
I would prefer not to introduce ifunc usage on aarch64. Our existing usages have already caused a lot of trouble for little gain.

dstogov replied:
Then, it's easier to keep the patch as is. I can't test it, but speed improvement looks promising.

sebpop commented:
I rebased the patch on master, and tested with `make test`.
I removed the alignment code from before the loop to reduce code size and no performance change.
@dstogov the patch is ready to be committed to master and php-7.4, thanks!

nikic commented on ext/standard/crc32.c:
```diff
@@ -35,6 +61,34 @@ PHP_NAMED_FUNCTION(php_if_crc32)
+		if (nr) {
+			p = (char *)p2;
+			crc = __crc32b(crc, *p);
+                        --nr;
```
nit: Stray indentation.

sebpop replied:
Thanks @nikic for the review.
I just pushed a corrected version with all three cases looking the same.

nikic commented on ext/standard/crc32.c:
```diff
@@ -35,6 +61,34 @@ PHP_NAMED_FUNCTION(php_if_crc32)
+			p2 = (int16_t *)p4;
+		} else {
+			p2 = (int16_t *)p8;
+		}
```
Maybe make the structure the same for all three padding cases? Something like this:

```
int32_t *p4; // add above

if (nr >= sizeof(int32_t)) {
    p4 = (int32_t *) p8;
    crc = __crc32w(crc, *p4++);
    nr -= sizeof(int32_t);
}
if (nr >= sizeof(int16_t)) {
    p2 = (int16_t *) p4;
    crc = __crc32h(crc, *p2++);
    nr -= sizeof(int16_t);
}
// last one as is
```

nikic commented:
Merged as https://github.com/php/php-src/commit/2a535a9707c89502df8bc0bd785f2e9192929422, thanks!
